### PR TITLE
make query caches use `Sharded` only for multiple threads

### PR DESCRIPTION
**Pull request by SparrowLii**
This PR make query caches use `Sharded` only for multiple threads. This can improve spatial locality in single-threaded scenarios to improve performance

cc @Zoxc
r? @cjgillot

### CODE REVIEW

File: compiler/rustc_middle/src/query/plumbing.rs [hunk 1]
```diff
@@ -382,19 +433,19 @@ macro_rules! define_callbacks {
             #[inline(always)]
             pub fn $name(self, key: query_helper_param_ty!($($K)*)) -> $V
             {
-                restore::<$V>(query_get_at(
+                restore::<$V>(with_query_caches!(query_get_at(
```
Comment (Zoxc) [inline compiler/rustc_middle/src/query/plumbing.rs:436]: This change is probably a bit costly as is doubles the size of the query accessor.

**Comment by Zoxc** [comment]
I did a benchmark run for this PR with `cfg(parallel_compiler)` and 1 thread with https://github.com/rust-lang/rust/pull/111713 and a local branch with optimized `lock_shard` methods applied:
<table><tr><td rowspan="2">Benchmark</td><td colspan="1"><b>Before</b></th><td colspan="2"><b>After</b></th></tr><tr><td align="right">Time</td><td align="right">Time</td><td align="right">%</th></tr><tr><td>🟣 <b>clap</b>:check</td><td align="right">1.6604s</td><td align="right">1.6786s</td><td align="right">💔  1.09%</td></tr><tr><td>🟣 <b>hyper</b>:check</td><td align="right">0.2419s</td><td align="right">0.2442s</td><td align="right"> 0.96%</td></tr><tr><td>🟣 <b>regex</b>:check</td><td align="right">0.9228s</td><td align="right">0.9339s</td><td align="right">💔  1.20%</td></tr><tr><td>🟣 <b>syn</b>:check</td><td align="right">1.5121s</td><td align="right">1.5293s</td><td align="right">💔  1.14%</td></tr><tr><td>🟣 <b>syntex_syntax</b>:check</td><td align="right">5.9729s</td><td align="right">6.0290s</td><td align="right"> 0.94%</td></tr><tr><td>Total</td><td align="right">10.3102s</td><td align="right">10.4150s</td><td align="right">💔  1.02%</td></tr><tr><td>Summary</td><td align="right">1.0000s</td><td align="right">1.0107s</td><td align="right">💔  1.07%</td></tr></table>

**Comment by SparrowLii** [comment]
Yea I don't know why rustc-perf makes this PR starves. I gonna try use `RefCell` directly.

**Comment by SparrowLii** [comment]
@Zoxc Could you help test the perf again? If the performace not as good as your local test, it is very good. It means we can fetch better performace than [previous work](https://perf.rust-lang.org/compare.html?start=77c836e1ae582661924d3b6ec4d57a2de120f59f&end=2354ebd53732be5d62fc44ea9c49b197e6a39792&stat=wall-time&nonRelevant=true&secondary=false)

File: compiler/rustc_middle/src/query/plumbing.rs (hunk 1 above)
Comment (SparrowLii) [reply compiler/rustc_middle/src/query/plumbing.rs]: Refers to the rustc-perf's result, this change doesn't cause regression?

Comment (Zoxc) [reply compiler/rustc_middle/src/query/plumbing.rs]: That doesn't really affect the code size.

**Comment by apiraino** [comment]
Switching to waiting on author to resolve conflicts. I'm reading [this comment](https://github.com/rust-lang/rust/pull/115162#issuecomment-1692932383): do the latest perf. run results align with your expectations? Feel free to request a review with `@rustbot ready` if you need another look at this, thanks! 

@rustbot author

**Comment by SparrowLii** [comment]
Close this since we already had better solutions

**Pull request closed without merging**